File: 221.maximal-square.cpp
/*
 * @lc app=leetcode id=221 lang=cpp
 *
 * [221] Maximal Square
 *
 * https://leetcode.com/problems/maximal-square/description/
 *
 * algorithms
 * Medium (35.65%)
 * Likes:    2439
 * Dislikes: 63
 * Total Accepted:    207.3K
 * Total Submissions: 575.5K
 * Testcase Example:
 * '[["1","0","1","0","0"],["1","0","1","1","1"],["1","1","1","1","1"],["1","0","0","1","0"]]'
 *
 * Given a 2D binary matrix filled with 0's and 1's, find the largest square
 * containing only 1's and return its area.
 *
 * Example:
 *
 *
 * Input:
 *
 * 1 0 1 0 0
 * 1 0 1 1 1
 * 1 1 1 1 1
 * 1 0 0 1 0
 *
 * Output: 4
 *
 */

#include <bits/stdc++.h>
using namespace std;

// @lc code=start
class Solution {
 public:
  int maximalSquare(vector<vector<char>> &matrix) {
    if (matrix.empty() or matrix.front().empty()) {
      return 0;
    }
    vector<int> top;
    const int N = size(matrix.front());
    top.reserve(N);
    for (const char &c : matrix.front()) {
      top.push_back(c == '1');
    }
    vector<int> dp = top;
    int max_edge_length = *max_element(cbegin(dp), cend(dp));
    for (auto b = cbegin(matrix) + 1, e = cend(matrix); b != e; ++b) {
      const vector<char> &row = *b;
      int left = row.front() == '1';
      int prev = dp.front();
      dp.front() = left;
      max_edge_length = max(max_edge_length, left);
      for (int j = 1; j < N; ++j) {
        int num = row[j] == '1';
        int cache = dp[j];
        dp[j] = num ? min({prev, left, top[j]}) + 1 : 0;
        max_edge_length = max(max_edge_length, dp[j]);
        prev = cache;
        left = num ? left + 1 : 0;
        top[j] = num ? top[j] + 1 : 0;
      }
    }
    return max_edge_length * max_edge_length;
  }
};
// @lc code=end

File: 221.maximal-square.py
#
# @lc app=leetcode id=221 lang=python3
#
# [221] Maximal Square
#
# https://leetcode.com/problems/maximal-square/description/
#
# algorithms
# Medium (35.65%)
# Likes:    2439
# Dislikes: 63
# Total Accepted:    207.3K
# Total Submissions: 575.5K
# Testcase Example:  '[["1","0","1","0","0"],["1","0","1","1","1"],["1","1","1","1","1"],["1","0","0","1","0"]]'
#
# Given a 2D binary matrix filled with 0's and 1's, find the largest square
# containing only 1's and return its area.
#
# Example:
#
#
# Input:
#
# 1 0 1 0 0
# 1 0 1 1 1
# 1 1 1 1 1
# 1 0 0 1 0
#
# Output: 4
#
#

from bisect import *
from collections import *
from functools import *
from heapq import *
from itertools import *
from math import *
from operator import *
from typing import *


# @lc code=start
class Solution:
    def maximalSquare(self, matrix: List[List[str]]) -> int:
        if not (matrix and matrix[0]):
            return 0
        top = list(map(int, matrix[0]))
        # The edge length of the maximal all '1' square
        # which its bottom-right corner is matrix[i][j].
        dp = top[:]
        max_edge_length = max(dp)
        for row in islice(matrix, 1, None):
            first = int(row[0])
            left = first
            prev = dp[0]
            dp[0] = first
            max_edge_length = max(max_edge_length, dp[0])
            for j, c in islice(enumerate(row), 1, None):
                num = int(c)
                cache = dp[j]
                dp[j] = num and min(prev, left, top[j]) + 1
                max_edge_length = max(max_edge_length, dp[j])
                prev = cache
                left = num and left + 1
                top[j] = num and top[j] + 1
        return max_edge_length * max_edge_length


# @lc code=end
